src/index.ts

```diff
@@ -1,17 +1,75 @@
+import * as fs from 'fs';
+import * as https from 'https';
+import * as path from 'path';
+
 const core = require('@actions/core');
 const exec = require('@actions/exec');
 
 import buildExec from './buildExec';
 
-const {execArgs, options, failCi} = buildExec();
-
-exec.exec('bash', execArgs, options)
-    .catch((err) => {
-      if (failCi) {
-        core.setFailed(
-            `Codecov failed with the following error: ${err.message}`,
-        );
-      } else {
-        core.warning(`Codecov warning: ${err.message}`);
-      }
-    });
+let failCi;
+const setFailure = (message, failCi) => {
+   failCi ? core.setFailed(message) : core.warning(message);
+};
+const isWindows = (platform) => {
+  return platform === 'windows';
+};
+const PLATFORMS = ['alpine', 'linux', 'macos', 'windows'];
+
+try {
+  const {execArgs, options, failCi, platform} = buildExec();
+  if (!PLATFORMS.includes(platform)) {
+    setFailure(
+        `Codecov: Encountered an unexpected platform: ${platform}`,
+        failCi,
+    );
+    process.exit();
+  }
+  const url = `https://uploader.codecov.io/latest/codecov-${platform}`;
+  const filename = path.join(
+      __dirname,
+      `uploader${isWindows(platform) ? '.exe' : ''}`,
+  );
+
+  https.get(url, (res) => {
+    // Image will be stored at this path
+    const filePath = fs.createWriteStream(filename);
+    res.pipe(filePath);
+    filePath
+        .on('error', (err) => {
+          setFailure(
+              `Codecov: Failed to write uploader binary: ${err.message}`,
+              failCi,
+          );
+        }).on('finish', () => {
+          filePath.close();
+          core.info('Uploader binary written.');
+
+          // TODO - validate step
+
+          fs.chmodSync(filename, '777');
+          exec.exec(filename, execArgs, options)
```

Comment:
This should be awaited via the `await` keyword. Two things to note:

- The enclosing function should be made async on line 44
- Lines 61 to 70 should be moved before line 51 to prevent a TypeError from being thrown

```suggestion
          await exec.exec(filename, execArgs, options)
```

Reply:
Nice, makes sense. For my future reference, how did you realize this was an issue?

Reply:
Good question.

I was able to discover this due to looking at the imports and realizing two were still brought in via `require()` instead of `import`. This led me to understand that the type definitions of those modules were not brought in and available to the developer. Once they were, it became directly apparent that `exec()` was typed to return a `Promise<number>`.

Reply:
An unfortunate side effect when dealing with promises is that it's quite easy for them to "dangle" if their settlement is never handled. TypeScript does not alert for this but Node.js will emit an error message indicating an UnhandledRejection has occurred if a dangling promise is rejected / throws an error.

Reply:
There's a TSLint rule ([no-floating-promises](https://palantir.github.io/tslint/rules/no-floating-promises/)) that checks for this behavior. ESLint does not have a built-in rule that corresponds with this behavior although there are a few from the community (e.g., the one from the typescript-eslint project) that handles it well.